Feature/pybackend

vhermecz opened the pull request:
Havent run the testsuite yet on it, still have a few TODO/NOTE/FIXME markers, but overall I've manually went through most of the features...

vhermecz committed:
Initial version of python impl

vhermecz committed:
initial doc

bobthecow commented on src/py/genghis/models.py:
```diff
@@ -0,0 +1,408 @@
+        if self.db is None:
+            return db_name in self.client.database_names()
+        else:
+            self.db = db_name
```
`==`?

vhermecz replied:
Oh good jesus, I was like WTF when I wrote that down, great thing you discovered it... :) Fixing

bobthecow commented on src/py/genghis/models.py:
```diff
@@ -0,0 +1,408 @@
+                json["error"] = "Authentication error: {e.message}".format(e=e)
+        return json
+    
+    def to_json(self):  # ?
```
The reason the server model has both #as_json and #to_json in the Ruby implementation is idiomatic. If it doesn't make sense here, there's no reason for it. If that's what you meant by your `?` comment :)

vhermecz replied:
Good point, will add it to my TODO list

bobthecow commented on src/py/genghis/models.py:
```diff
@@ -0,0 +1,408 @@
+            self._info = db
+        return self._info
+            
+    @property
```
For #info and #stats, only make them a `@property` if it makes sense in Python. Ruby treats _everything_ like that, so it's idiomatic Ruby to use the "property-ish" version when a method has no arguments. No need to write weird Python just to match a Ruby idiom :)

bobthecow replied:
This applies across the board, for what it's worth. The Ruby and PHP implementations have identical external APIs, but are implemented idiomatically for their respective languages.

bobthecow commented:
Great start so far!

Let's target merging this with the [three-dot-oh-my branch](https://github.com/bobthecow/genghis/tree/feature/three-dot-oh-my), which is where development for the 3.0 release is happening. The assets are structured slightly differently, and the API has one or two minor changes, but you should be able to cherry pick this branch into a branch made from that branch without issue.

vhermecz committed:
x fix src/py/genghis/models.py - Ugly error
c chg src/py/readme.md - Few more TODOs

vhermecz committed:
c chg src/py/genghis/helpers.py - make_response mimetype fix, removed unused json methods
c chg src/py/genghis/models.py - Added note on non utf-8 os message
c chg src/py/genghis/server.py - make_response mimetype fix, method renaming
c chg src/py/readme.md - Dependencies provided

lunixbochs commented on src/py/genghis/helpers.py:
```diff
@@ -0,0 +1,111 @@
+def request_json():
+    try:
+        return decode(request.data)
+    except:
```
You almost never want `except:` by itself. That hides weird stuff like keyboard interrupts and memory errors.

Use `except Exception:` instead.

lunixbochs commented on src/py/genghis/models.py:
```diff
@@ -0,0 +1,409 @@
+                self.db = database
+            
+            self.name = name
+        except InvalidURI:
```
This `try` block is probably way too big. It looks like only `_extract_extra_options` on line 35 might throw InvalidURI, then the `except` basically acts as a goto.

lunixbochs commented on src/py/genghis/models.py:
```diff
@@ -0,0 +1,409 @@
+        return json.dumps(self.as_json())
+    
+    @property
+    def pages(self):
```
Making everything an `@property` can hide the fact a query happens on access.

No need to make class methods into properties just because they have no arguments. This is probably a pattern you'd see more in Ruby than Python.

vhermecz committed:
Correcting ugly exc handling

bobthecow commented:
The current merge target should be the develop branch, since feature/three-dot-oh-my has been merged there.

vhermecz committed:
Adding changes from feature/three-dot-oh-my branch
c chg src/py/genghis/jsonutil.py - Adding timestamp and NaN
c chg src/py/genghis/models.py - Adding truncate and explain
c chg src/py/genghis/server.py - Updating location of assets, name=>url param change, truncate/explain methods, mime change for file request

vhermecz committed:
Adding projection and sorting for document retrieval
c chg src/py/genghis/helpers.py - Adding fields and sort req param accessors
c chg src/py/genghis/models.py - Adding fields and sort to Query, and Collection.documents
c chg src/py/genghis/server.py - Adding fields and sort to documents API endpoint
